### ISSUE TITLE
Core/Array: reset

### ISSUE THREAD

**Issue by nkkollaw** [nkkollaw/zubr#81]

**Comment by Petah** [comment]
Is this even possible to implement as an alias?

```
<?php
$a = [1, 2, 3];

var_dump(next($a));
var_dump(next($a));
var_dump(reset($a));
var_dump(next($a));

function test_reset(&$array) {
    reset($array);
}

var_dump(next($a));
var_dump(next($a));
var_dump(test_reset($a));
var_dump(next($a));
```

Output:

```
int(2)
int(3)
int(1)
int(2)
int(3)
bool(false)
NULL
int(2)
```

https://3v4l.org/por5c

**Comment by nkkollaw** [comment]
As discussed, this function needs its parameter to be passed by reference.

What to do, considering that we're trying to avoid that? I think this is a gotcha but we'd have to do by reference.

**Comment by ddziaduch** [comment]
@nkkollaw I don't see any other option that sticking to the reference

**Comment by radmen** [comment]
> What to do, considering that we're trying to avoid that? I think this is a gotcha but we'd have to do by reference.

The question is more general - should the wrapper function be pure? If so it means that some native functions (like `reset()`, `key()`, `current()`) will be broken "by design".

If the answer is *no* - I wouldn't bother with passing variables as references.

**Comment by nkkollaw** [comment]
In this case, the only purpose of the functions is to modify the element passed, so it kind of makes sense. 

Implementing without passing by reference wouldn't work IMO.

I would definitely add it to the documentation as a gotcha, but I think we should make ot work like the original.

**Comment by Petah** [comment]
> the only purpose of the functions is to modify the element passed 

Its not always the only purpose. For example `end` also returns the last element of the array (as well as setting the pointer to the end).

**Comment by nkkollaw** [comment]
Definitely. I meant that if you take that aspect away, the function doesn't make much sense.
